Sources/FoundationInternationalization/ICU/ICU+StringConverter.swift

```diff
@@ -0,0 +1,202 @@
+            }
+
+            let srcLength = uchars.count
+            let capacity = srcLength * Int(ucnv_getMaxCharSize(converter)) + 1
+            let dest = UnsafeMutableRawPointer.allocate(
+                byteCount: capacity,
+                alignment: MemoryLayout<CChar>.alignment
+            )
+
+            var status: UErrorCode = U_ZERO_ERROR
+            if lossy {
+                var lossyChar: UChar = encoding == .ascii ? 0xFF : 0x3F
+                ucnv_setSubstString(
+                    converter,
+                    &lossyChar,
+                    1,
+                    &status
+                )
+                guard status.isSuccess else { return nil }
+
+                ucnv_setFromUCallBack(
+                    converter,
+                    UCNV_FROM_U_CALLBACK_SUBSTITUTE,
+                    nil, // newContext
+                    nil, // oldAction
+                    nil, // oldContext
+                    &status
+                )
+                guard status.isSuccess else { return nil }
+            } else {
+                ucnv_setFromUCallBack(
+                    converter,
+                    UCNV_FROM_U_CALLBACK_STOP,
+                    nil, // newContext
+                    nil, // oldAction
+                    nil, // oldContext
+                    &status
+                )
+                guard status.isSuccess else { return nil }
+            }
+
+            let actualLength = ucnv_fromUChars(
+                converter,
+                dest,
+                CInt(capacity),
+                uchars.baseAddress,
+                CInt(srcLength),
+                &status
+            )
+            guard status.isSuccess else { return nil }
+            return Data(
+                bytesNoCopy: dest,
+                count: Int(actualLength),
+                deallocator: .custom({ pointer, _ in pointer.deallocate() })
+            )
+        }
+    }
+}
+
+extension ICU.StringConverter {
+    nonisolated(unsafe) static private var _converters: LockedState<[String.Encoding: ICU.StringConverter]> = .init(initialState: [:])
+
+    static func converter(for encoding: String.Encoding) -> ICU.StringConverter? {
+        return _converters.withLock {
+            if let converter = $0[encoding] {
+                return converter
+            }
+            if let converter = ICU.StringConverter(encoding: encoding) {
+                $0[encoding] = converter
+                return converter
+            }
+            return nil
+        }
+    }
+}
+
+
+@_dynamicReplacement(for: _icuMakeStringFromBytes(_:encoding:))
+func _icuMakeStringFromBytes_impl(_ bytes: UnsafeBufferPointer<UInt8>, encoding: String.Encoding) -> String? {
+    guard let converter = ICU.StringConverter.converter(for: encoding),
+          let pointer = bytes.baseAddress else {
+        return nil
+    }
+    let data =  Data(
+        bytesNoCopy: UnsafeMutableRawPointer(mutating: pointer),
```

Comment:
Is it worth adding a comment here about this being "safe" because this data neither escapes this function nor is mutated? Since the pointer provided is not guaranteed to live longer than the function or be mutable, this could be a trap for a later addition that doesn't see those two constraints

Reply:
I agree it's better to have some comments. I'm going to add ones.